### PR TITLE
Adds UVCal get_* and utils.uvcalibrate

### LINKED ISSUE #242: function for grabbing gain of ant/pol
Similar to UVData.get_data, we should have a function for quickly grabbing the gain for an antenna/pol.

### PR DESCRIPTION
**Pull request by nkern**
Adds `UVCal.get_` methods and `utils.uvcalibrate`

## Description

Adds slicing and get methods for `UVCal` and `utils.uvcalibrate` which automatically calibrates a `UVData` given a `UVCal`

## Motivation and Context
This helps streamline analysis with UVCal

closes #242 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass in both python 2 and python 3.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,8 @@ All notable changes to this project will be documented in this file.
 - `UVData.conjugate_bls` method to conjugate baselines to get the desired baseline directions.
 - `UVData.reorder_blts` method to reorder the data along the blt axis (and optionally also conjugate baselines), and a new `blt_order` optional parameter on UVData objects to track the ordering (including through read/writes).
 - `lst_array` is now saved to UVFITS files (even though it's not a standard parameter) so that it doesn't have to be recalculated
+- `UVCal.get_*` methods for accessing data arrays with antenna-polarization keys
+- `utils.uvcalibrate` for automated calibration of `UVData` by `UVCal`
```
Comment (bhazelton) [inline CHANGELOG.md]: These should go under the Unreleased section

File: pyuvdata/utils.py [hunk 2]
```diff
@@ -388,6 +388,59 @@ def unphase_uvw(ra, dec, uvw):
...
+        pol_ind = np.argmin(np.abs(uvdata.polarization_array - polstr2num(key[2])))
+
+        # try to get gains for each antenna
+        ant1 = (key[0], key[2][0])
+        ant2 = (key[1], key[2][1])
+        if not uvcal._has_key(*ant1) or not uvcal._has_key(*ant2):
+            if flag_missing:
+                uvdata.flag_array[blt_inds, 0, :, pol_ind] = True
+            continue
+        gain = (uvcal.get_gains(ant1) * np.conj(uvcal.get_gains(ant2))).T  # tranpose to match uvdata shape
```
Comment (bhazelton) [inline pyuvdata/utils.py]: Does this mean that this function won't work for a delay calibration? If so, there should be a fail-fast error raised that says so.

**Comment by bhazelton** [comment]
@adampbeardsley can you take a look at this PR? I know you worked a bit on an apply_cal method in the past so I think you're more in touch with the subtleties there.

File: pyuvdata/utils.py [hunk 3]
```diff
@@ -388,6 +388,59 @@ def unphase_uvw(ra, dec, uvw):
...
+        uvcal: UVCal object
+        inplace: bool, if True edit uvdata in place, else deepcopy
+        prop_flags : bool, if True, propagate calibration flags to data flags
+            and doesn't use flagged gains. Otherwise, uses flagged gains and
+            does not propagate calibration flags to data flags.
+        flag_missing: bool, if True, flag baselines in uvdata
+            if a participating antenna or polarization is missing in uvcal.
+
+    Returns:
+        uvdata: If not inplace, return deepcopied uvdata
```
Comment (bhazelton) [inline pyuvdata/utils.py]: Please put this and all the docstrings into the numpy-style (https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)

Comment (nkern) [reply pyuvdata/utils.py]: This this a recent change? It seems like all doc-strings in pyuvdata do not follow numpy style, or am I confused?

Comment (bhazelton) [reply pyuvdata/utils.py]: We are in the process of converting, so we ask that PRs that touch existing docstrings or add new ones are in the numpy style. You'll notice that one of the PR tickboxes addresses this.

File: docs/tutorial.rst [hunk 4]
```diff
@@ -1112,6 +1112,53 @@ b) FHD cal to cal fits
...
+  >>> print(gain.shape)
+  (1024, 3, 2)
+  >>> print(UVC.Nfreqs, UVC.Ntimes, UVC.Njones)
+  (1024, 3, 2)
+
+  # One can also request flags and quality arrays in a similar manner
+  >>> flags = UVC.get_flags(9, 'Jxx')
+  >>> quals = UVC.get_quality(9, 'Jxx')
+
+b) Calibration of UVData by UVCal
```
Comment (bhazelton) [inline docs/tutorial.rst]: I think this should get it's own larger heading, it's a big deal and doesn't really fall under "quick data access".

File: pyuvdata/utils.py [hunk 5]
```diff
@@ -392,23 +392,33 @@ def uvcalibrate(uvdata, uvcal, inplace=True, prop_flags=True, flag_missing=True)
...
+    flag_missing: bool, optional
+        if True, flag baselines in uvdata
+        if a participating antenna or polarization is missing in uvcal.
 
-    Returns:
-        uvdata: If not inplace, return deepcopied uvdata
+    Returns
+    -------
+    UVData
+        Retruns if not inplace
```
Comment (bhazelton) [inline pyuvdata/utils.py]: Retruns -> Returns. Also you can note that this is an optional return by adding `, optional` after `UVData`

File: pyuvdata/uvcal.py [hunk 6]
```diff
@@ -691,11 +691,15 @@ def ant2ind(self, antnum):
...
-        Returns:
-            index : Index in data arrays (integer)
+        Parameters
+        ----------
+        antnum : int
+            Antenna number
+
+        Returns
+        -------
+        intn
```
Comment (bhazelton) [inline pyuvdata/uvcal.py]: I think this should be `int` not `intn`

File: pyuvdata/utils.py [hunk 7]
```diff
@@ -388,6 +388,69 @@ def unphase_uvw(ra, dec, uvw):
...
+    """
+    # deepcopy for not inplace
+    if not inplace:
+        uvdata = copy.deepcopy(uvdata)
+
+    # input checks
+    if uvcal.cal_type != 'gain':
+        raise ValueError("uvcal must have cal_type of 'gain' for calibration")
```
Comment (adampbeardsley) [inline pyuvdata/utils.py]: As a user, if I had a delay type UVCal object, I would assume the calibrate function would convert to gain and apply (even if it made a copy of the UVCal object to not disturb the original).

File: pyuvdata/uvcal.py [hunk 8]
```diff
@@ -661,6 +661,103 @@ def convert_to_gain(self, delay_convention='minus', run_check=True, check_extra=
         else:
             raise ValueError('cal_type is unknown, cannot convert to gain')
 
+    def get_gains(self, *args):
```
Comment (adampbeardsley) [inline pyuvdata/uvcal.py]: The args here should be documented - what should they be?
We also are trying to stay away from args and kwargs when possible. Bryna did a really nice job of revamping the `get_X` methods on UVData - we should try to match that here.

**Comment by nkern** [comment]
@adampbeardsley thanks for the suggestions. Here is my stab at incorporating your d-term comments.

File: pyuvdata/tests/test_utils.py [hunk 9]
```diff
@@ -930,3 +931,47 @@ def test_and_collapse_weights():
...
+
+    # division calibrate
+    key = (43, 72, 'xx')
+    ant1 = (43, 'Jxx')
+    ant2 = (72, 'Jxx')
+    uvc.gain_convention = 'divide'
+    uvdcal = uvutils.uvcalibrate(uvd, uvc, prop_flags=False, flag_missing=False, inplace=False)
+    np.testing.assert_array_almost_equal(uvdcal.get_data(key), uvd.get_data(key) / (uvc.get_gains(ant1) * uvc.get_gains(ant2).conj()).T)
+
+    # multiplication calibrate
```
[the remark this answers is not in the document]
Comment (nkern) [reply pyuvdata/tests/test_utils.py]: Okay, I think it already strikes a good balance between scope and speed, but I'll chunk it at finer level and you can tell me what you think.

**Comment by adampbeardsley** [comment]
Thanks for addressing my comments @nkern! Looks good to me - except the tutorial errors you're getting :)

**Pull request merged**